### PR TITLE
ch4/fbox: uninline fastbox initialization code

### PR DESCRIPTION
**Pull request by gcongiu**
## Pull Request Description

Uninline fastbox init code

## Expected Impact
No performance impact as this is non-critical path

## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [x] Commits are self-contained and do not do two things at once
* [x] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [ ] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@wesbland I had a discussion with @raffenet about the different options for uninlining fastboxes init code. One of the options that came out was to get rid of the function table, which allows for runtime selection of the eager protocol (either fbox or iqueue). Since we would like to support both (rather than either one or the other) I was suggesting we would go for this option. Do folks at Intel have any strong opposition/motivation against this?

**Comment by wesbland** [comment]
> @wesbland I had a discussion with @raffenet about the different options for uninlining fastboxes init code. One of the options that came out was to get rid of the function table that allows for runtime selection of the eager protocol (either fbox or iqueue). Since we would like to support both (rather than either one or the other) I was suggesting we would go for this option. Do folks at Intel have any strong opposition/motivation against this? Or prefer a different solution instead?

I don't have a strong objection I guess. I don't use the runtime switching feature myself. I don't know if @ddurnov has another opinion. We both use iqueue (#3873) full time I think.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by ddurnov** [comment]
@gcongiu @raffenet We use eager modules infrastructure very extensively (39 modules total) We use it for different platforms. 
Initial idea behind the infrastructure was similar to netmods infra. E.g. in case you have only one module, then there won't be any indirection. In case of several modules (each in its own folder) we use the function table.

**Comment by raffenet** [comment]
@ddurnov ok, thanks for confirming this is in use. We should ensure the infrastructure is complete in MPICH and put in guards (Jenkins tests) to prevent it from breaking.

**Comment by gcongiu** [comment]
@raffenet it seems we will need to keep the current selection infrastructure merging `fbox` and `iqueue` in a single posix module to still allow runtime selection.

**Comment by raffenet** [comment]
test:jenkins/ch4/most

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@minsii testing ofi. If tests finish correctly can you start a review of this PR?

File: src/mpid/ch4/shm/posix/eager/stub/stub_init.h [hunk 1]
```diff
@@ -13,13 +13,13 @@
 
 #include "stub_impl.h"
 
-MPL_STATIC_INLINE_PREFIX int MPIDI_POSIX_eager_init(int rank, int size)
+int MPIDI_POSIX_eager_init(int rank, int size)
 {
```
Comment (minsii) [inline src/mpid/ch4/shm/posix/eager/stub/stub_init.h]: Changes for `stub_init` should be in a separate commit.

File: src/mpid/ch4/shm/posix/eager/fbox/posix_eager_inline.h [hunk 2]
```diff
@@ -11,8 +11,9 @@
 #ifndef POSIX_EAGER_FBOX_DIRECT_H_INCLUDED
 #define POSIX_EAGER_FBOX_DIRECT_H_INCLUDED
 
-#include "fbox_noinline.h"
 #include "fbox_send.h"
 #include "fbox_recv.h"
 
+#include "fbox_noinline.h"
+
 #endif /* POSIX_EAGER_FBOX_DIRECT_H_INCLUDED */
```
Comment (minsii) [inline src/mpid/ch4/shm/posix/eager/fbox/posix_eager_inline.h]: This commit need cleanup. This change is unneeded.

File: src/mpid/ch4/shm/posix/subconfigure.m4 [hunk 3]
```diff
@@ -93,7 +107,9 @@ MPIDI_POSIX_eager_${posix_eager}_recv_transaction_t ${posix_eager};"
     AC_SUBST(ch4_posix_eager_array_sz)
     AC_SUBST(ch4_posix_eager_strings)
     AC_SUBST(ch4_posix_eager_func_decl)
+    AC_SUBST(ch4_posix_eager_native_func_decl)
     AC_SUBST(ch4_posix_eager_func_array)
+    AC_SUBST(ch4_posix_eager_native_func_array)
     AC_SUBST(ch4_posix_eager_pre_include)
```
Comment (minsii) [inline src/mpid/ch4/shm/posix/subconfigure.m4]: Why do you add `MPIDI_POSIX_eager_native_funcs*`?

File: src/mpid/ch4/shm/posix/eager/stub/stub_init.h (hunk 1 above)
Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/stub/stub_init.h]: • I have used the netmod as blueprint for the uninline. If we don't move `init|finalize` to a separate .c file these will be undefined at link time when inline is disabled;
• To split the `init|finalize` functions from the rest of the inlined functions as noted in previous point.

There might be a better way of doing this. I have tried to follow the netmod for consistency within the code.

**Comment by gcongiu** [comment]
@minsi I have tried using the netmod as blueprint to do the uninline both for simplicity and also for consistency with the rest of the code. More specifically to your comment, IIRC leaving `init|finalize` in the same `posix_eager_impl.h` would cause these to have duplicate symbols at linking time; thus also the need for splitting the function tables into `funcs` and `native_funcs`. There might be a better way of doing the uninline but I need to double check.

File: src/mpid/ch4/shm/posix/eager/stub/stub_init.h (hunk 1 above)
Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/stub/stub_init.h]: Should this commit change .h to .c?

File: src/mpid/ch4/shm/posix/eager/src/posix_eager_impl.c [hunk 4]
```diff
@@ -0,0 +1,24 @@
+/*
+ *  (C) 2020 by Argonne National Laboratory.
+ *      See COPYRIGHT in top-level directory.
+ *
+ */
+
+#ifndef POSIX_EAGER_INLINE
+#ifndef POSIX_EAGER_DISABLE_INLINES
+
```
Comment (minsii) [inline src/mpid/ch4/shm/posix/eager/src/posix_eager_impl.c]: I do not think the `#ifndef POSIX_EAGER_INLINE` wrapper is needed for .c

File: src/mpid/ch4/shm/posix/eager/fbox/fbox_init.c [hunk 5]
```diff
@@ -30,7 +30,7 @@
 
 #define MPIDI_POSIX_MAILBOX_INDEX(sender, receiver) ((MPIDI_POSIX_global.num_local) * (sender) + (receiver))
 
-int MPIDI_POSIX_eager_init(int rank, int size)
+int MPIDI_POSIX_fbox_init(int rank, int size)
 {
```
Comment (minsii) [inline src/mpid/ch4/shm/posix/eager/fbox/fbox_init.c]: The commit message does not explain why you need change the func name. I like the renaming. But perhaps it should not be in this commit. How about move it to the fbox uninline commit? Do you need the same renaming for stub?

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/fbox/fbox_init.c]: I will move the name change and fix the commit message. Concerning stub, as I understand it these are there only as blueprint for developers that want to add new modules; So probably I should also change these although it does not affect functionality.

File: src/mpid/ch4/shm/posix/eager/src/posix_eager_impl.c (hunk 4 above)
Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/src/posix_eager_impl.c]: Then the code would be always compiled? I am not even sure why we have the `POSIX_EAGER_DISABLE_INLINES` in `posix_eager_impl.c` (and the netmod as well) as the define is done in a separate compilation unit (i.e. `fbox/func_table.c`).

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/src/posix_eager_impl.c]: As long as posix/eager is built, this part has to be always compiled, right? I think we always build posix/eager.

It looks like regular wrapper for header file, but should not be used for .c. Can you please test without the macros?

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/src/posix_eager_impl.c]: I think `POSIX_EAGER_DISABLE_INLINES` can be safely removed (in fact it is never defined in this compilation unit), `POSIX_EAGER_INLINE` should be left or else duplicated symbols are found when compiling with inlining enabled (I have tested it).

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/shm/posix/eager/fbox/func_table.c [hunk 6]
```diff
@@ -8,18 +8,29 @@
...
 #define POSIX_EAGER_DISABLE_INLINES
 
 #include <mpidimpl.h>
 #include <posix_eager.h>
 #include "../../posix_types.h"
-#include "posix_eager_inline.h"
+#include "fbox_noinline.h"
 
-MPIDI_POSIX_eager_funcs_t MPIDI_POSIX_eager_fbox_funcs = {
+MPIDI_POSIX_eager_init_funcs_t MPIDI_POSIX_eager_fbox_init_funcs = {
```
Comment (raffenet) [inline src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: This isn't going to fix multi-shmmod builds, right? This seems like just a hack to un-inline some init code.

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: Do you mean multi-eager protocols?

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: I have only tested fastbox for now. In principle this should work. The only difference is that besides unlinening the fbox code we also move to function tables for both inline and noinline cases for the init/finalize functions.

Comment (raffenet) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: > Do you mean multi-eager protocols?

Sorry, yes, I meant multi-eager builds. Though now that I try it, I am able to build both fbox and stub at the same time with this PR... Maybe I am misunderstanding the infrastructure. I will have to look at on Monday.

Comment (raffenet) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: Is this additional of another function table actually necessary here? Isn't the namespace and macro defines for inline configurations sufficient to allow this code to move out of inline functions?

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: As @minsii noted there is little benefit in avoiding function tables for `init|finalize`. Thus this PR also defaults to function tables for these two regardless of whether inlinining is enabled or not, this is why we need separate function tables. If you remove the last two commits then you should be left with what you are suggesting.

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: The benefit is that code is a little simpler as we don't have to use some trick to redefine init and finalize symbols when inlining is enabled and use guards in some part of the code.

Comment (raffenet) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: But we already use that trick for all the other functions in the eager modules, no?

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: We have only one eager module: `fbox`, now we are adding `iqueue`. I think it is easy to switch to a simpler approach for init/finalize.

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: For any other perf-critical functions, we should still use the trick.

Comment (raffenet) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: I'm disagreeing that an additional function table is simpler. I get init/finalize is off the critical path, but we do not make this differentiation anywhere else in ch4. This is more complex, IMO.

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: I do not have strong preference on always-function-table v.s. trick for init/finalize. We can switch back to the trick version if you think it is good for consistency.

Comment (gcongiu) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: Should I get rid of the last two commits then?

Comment (minsii) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: Yes, please.

Comment (raffenet) [reply src/mpid/ch4/shm/posix/eager/fbox/func_table.c]: New version looks good to me. I'm still disturbed by the differences in this infrastructure vs. the netmods, but I don't think that needs to be solved right this moment.

**Comment by gcongiu** [comment]
test:jenkins/ch4/ofi

**Comment by gcongiu** [comment]
@minsii @raffenet are you happy with this being merged?

**Comment by minsii** [comment]
Looks good to me. Approved.
